File: leetcode/medium/57. Insert Interval.ts
// need more cleaning
function insert(intervals: number[][], newInterval: number[]): number[][] {
  const overlap_intervals_idx: number[] = [];
  const result: number[][] = [...intervals];
  for (let i = 0; i < intervals.length; i += 1) {
    if (
      newInterval[0] <= intervals[i][1] &&
      newInterval[1] >= intervals[i][1]
    ) {
      overlap_intervals_idx.push(i);
      continue;
    }
    if (newInterval[1] >= intervals[i][0] && newInterval[0] < intervals[i][0]) {
      overlap_intervals_idx.push(i);
      continue;
    }
    if (
      newInterval[0] >= intervals[i][0] &&
      newInterval[1] <= intervals[i][1]
    ) {
      overlap_intervals_idx.push(i);
      continue;
    }
  }
  if (overlap_intervals_idx.length === 0) {
    // if there are/is no overlap interval
    if (
      (intervals[0] && newInterval[1] < intervals[0][0]) ||
      result.length === 0
    ) {
      result.unshift(newInterval);
      return result;
    }
    for (let i = 0; i < result.length; i += 1) {
      if (
        i === result.length - 1 ||
        (newInterval[0] > result[i][1] && newInterval[1] < result[i + 1][0])
      ) {
        result.splice(i + 1, 0, newInterval);
        break;
      }
    }
    return result;
  }
  const new_interval: number[] = new Array(2);
  // fiding the smallest replacemet for the newe interval ,,, starting index [0]
  if (newInterval[0] > intervals[overlap_intervals_idx[0]][0]) {
    new_interval[0] = intervals[overlap_intervals_idx[0]][0];
  } else {
    new_interval[0] = newInterval[0];
  }
  // finding the biggest replacement for the new interval ,,,, ending index [1]
  if (
    newInterval[1] <
    intervals[overlap_intervals_idx[overlap_intervals_idx.length - 1]][1]
  ) {
    new_interval[1] =
      intervals[overlap_intervals_idx[overlap_intervals_idx.length - 1]][1];
  } else {
    new_interval[1] = newInterval[1];
  }

  // replacing the old interval to the new interval
  result.splice(
    overlap_intervals_idx[0],
    overlap_intervals_idx.length,
    new_interval
  );

  return result;
}

File: leetcode/medium/57. Insert Interval.rs
impl Solution {
    pub fn insert(intervals: Vec<Vec<i32>>, new_interval: Vec<i32>) -> Vec<Vec<i32>> {
        let mut new: Vec<Vec<i32>> = vec![];
        if intervals.len() == 0 {
            new.push(new_interval.to_vec());
            return new;
        }
        let mut included: Vec<usize> = vec![];

        for i in 0..intervals.len() {
            if intervals[i][0] <= new_interval[0] && intervals[i][1] >= new_interval[0] {
                included.push(i);
                continue;
            }
            if intervals[i][0] <= new_interval[1] && intervals[i][1] >= new_interval[1] {
                included.push(i);
                continue;
            }
            if new_interval[0] <= intervals[i][0] && new_interval[1] >= intervals[i][0] {
                included.push(i);
                continue;
            }
            if new_interval[0] <= intervals[i][1] && new_interval[1] >= intervals[i][1] {
                included.push(i);
                continue;
            }
            new.push(intervals[i].to_vec());
        }
        // println!("{:?}", included);
        if included.len() > 0 {
            let mut new_vec: Vec<i32> = vec![];
            new_vec.push(if intervals[included[0]][0] > new_interval[0] {
                new_interval[0]
            } else {
                intervals[included[0]][0]
            });
            new_vec.push(
                if intervals[included[included.len() - 1]][1] < new_interval[1] {
                    new_interval[1]
                } else {
                    intervals[included[included.len() - 1]][1]
                },
            );
            // println!("{:?}", new);
            new.insert(included[0], new_vec);
        } else if new_interval[0] < intervals[0][0]
            && new_interval[0] < intervals[0][1]
            && new_interval[1] < intervals[0][0]
            && new_interval[1] < intervals[0][1]
        {
            new.insert(0, new_interval.to_vec());
        } else if new_interval[0] > intervals[intervals.len() - 1][0]
            && new_interval[0] > intervals[intervals.len() - 1][1]
            && new_interval[1] > intervals[intervals.len() - 1][0]
            && new_interval[1] > intervals[intervals.len() - 1][1]
        {
            new.push(new_interval.to_vec());
        } else {
            for i in 0..intervals.len() {
                if new_interval[0] < intervals[i][0]
                    && new_interval[0] < intervals[i][1]
                    && new_interval[1] < intervals[i][0]
                    && new_interval[1] < intervals[i][1]
                {
                    new.insert(i, new_interval.to_vec());
                    break;
                }
            }
        }
        new
    }
}
